SNI exposition: user can control advertised broker port

robobario opened the pull request:
### Type of change

- Enhancement / new feature

### Description

1. Introduces `advertisedBrokerAddressPattern` to the SNI configuration.
2. Deprecates `brokerAddressPattern` in SNI configuration.
3. `advertisedBrokerAddressPattern` or `brokerAddressPattern` must be specified, but not both
4. `brokerAddressPattern` and `advertisedBrokerAddressPattern` can optionally end with an advertised port specifier `:[1-9][0-9]*$`. If specified, this port is the port returned to clients when we inject broker addresses into Kafka RPCs using the `BrokerAddressFilter`. It may differ from the port the VirtualCluster listens on.

Example configuration where we want kroxylicious to listen on 9092, but tell clients to connect to 443:

```
        clusterNetworkAddressConfigProvider:
          type: SniRoutingClusterNetworkAddressConfigProvider
          config:
            bootstrapAddress: mycluster-proxy.kafka:9092
            advertisedBrokerAddressPattern: broker$(nodeId).mycluster-proxy.kafka:443
```

Leaves documentation change for a separate PR

Closes #1760 

### Additional Context

It may be convenient to put Kroxylicious behind yet-another-proxy that uses a different port scheme. For example OpenShift Routes for TLS passthrough will all be exposed on port 443 by default. Having to align the proxy with this port scheme is a bit of a pain as we won't be running as root in container, so can't bind to the privileged low port.

Instead, we can give users control over the advertised port, so that Kroxylicious can present the brokers as being available on the client-facing port.

This is related to Kroxylicious Operator as we are wondering whether we should integrate directly with ingress routers.

### Checklist

- [ ] PR raised from a fork of this repository and made from a branch rather than main. 
- [ ] Write tests
- [ ] Update documentation
- [ ] Make sure all unit/integration tests pass
- [ ] Make sure all Sonarcloud warnings are addressed or are justifiably ignored.
- [ ] If applicable to the change, [trigger the system test suite](../blob/main/DEV_GUIDE.md#jenkins-pipeline-for-system-tests).  Make sure tests pass.
- [ ] If applicable to the change, [trigger the performance test suite](../blob/main/PERFORMANCE.md#jenkins-pipeline-for-performance). Ensure that any degradations to performance numbers are understood and justified.
- [ ] Ensure the PR references relevant issue(s) so they are closed on merging.
- [ ] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

> **_NOTE:_**  You must be a member of `@kroxylicious/developers` to trigger the system test and performance test suites.  If you are not part of this group, comment on the PR requesting a trigger, tagging `@kroxylicious/developers`.

robobario commented:
In the spirit of keeping changes small I haven't touched `bootstrapAddress`. I think it makes sense as you said @k-wall  to pull the listen port out and at the same time give users a way to configure the listen host. But can do that as another PR. There's #1278 covering this also.

k-wall commented on kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/ExpositionIT.java:
```diff
@@ -143,6 +144,60 @@ void exposesTwoClusterOverPlainWithSeparatePorts(KafkaCluster cluster) {
+                        SslConfigs.SSL_TRUSTSTORE_LOCATION_CONFIG, keystoreTrustStorePair.clientTrustStore(),
+                        SslConfigs.SSL_TRUSTSTORE_PASSWORD_CONFIG, keystoreTrustStorePair.password()))) {
+                    // do some work to ensure virtual cluster is operational
+                    createTopic(admin, TOPIC, 1);
```
I think it is worth describing the cluster and verifying that the ports are truely proxy.getLocalPort().

robobario replied:
Yep, at one point this test was passing, but the brokers were still being exposed as 9092. Will add

robobario replied:
added the check:

```
admin.describeCluster().nodes().get(5, TimeUnit.SECONDS).forEach(node -> {
                        assertThat(node.port()).isEqualTo(proxy.getLocalPort());
});
```

k-wall commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/clusternetworkaddressconfigprovider/SniRoutingClusterNetworkAddressConfigProvider.java:
```diff
@@ -113,41 +124,77 @@ public static class SniRoutingClusterNetworkAddressConfigProviderConfig {
 
         public SniRoutingClusterNetworkAddressConfigProviderConfig(@JsonProperty(required = true) HostPort bootstrapAddress,
-                                                                   @JsonProperty(required = true) String brokerAddressPattern) {
+                                                                   @Deprecated(forRemoval = true, since = "0.10.0") @JsonProperty String brokerAddressPattern,
```
You could just use a JsonAlias on advertisedBrokerAddressPattern to retain support for the old name.  It means brokerAddressPattern would start to accept a port, but that's a backwardly compatibility change which won't break existing use-cases.

robobario replied:
`brokerAddressPattern` is already updated to accept a port in this PR. I had the same thought, it's a backwards-compatible change and easier to code if there's no behaviour difference.

This way is uglier, but it gives us a couple of things:

1. YAML serialize/deserialize fidelity, which I think we lean on for tests and now in the operator we're using these classes to generate broker config.
2. The deprecated annotation, helps generate warnings and find the thing to kill off later. A bit more tooling support than putting a comment by the alias.

robobario replied:
ah, and a chance to log a deprecation warning

k-wall reviewed (commented):
Looks like a useful improvement and the change is going in the right direction.  Thanks @robobario

robobario commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/clusternetworkaddressconfigprovider/SniRoutingClusterNetworkAddressConfigProvider.java:
```diff
@@ -76,6 +82,11 @@ public HostPort getBrokerAddress(int nodeId) {
 
+        @Override
+        public int getAdvertisedPort(int nodeId) {
+            return advertisedPort;
```
1. sounds good to me.

robobario replied:
after having a play I think a. is okay for this PR. 

b) is appealing to me because the established `brokerAddress` we pass around is currently an odd combination of `advertisedHost` and `bindPort`, and the distinction between brokerAddress and advertisedBrokerAddress isn't immediately clear. However, the refactor is bigger and I'd tackle it in another PR.

robobario commented:
Have pushed up further changes to:
1. kill off usages of the deprecated props in kubernetes-examples
2. update `exposesTwoSeparateUpstreamClustersUsingSniRouting` SNI exposition test to be parameterized to run with old and new properties, to get some coverage on the old property. Updated the remaining tests to use the new property
3. Updated `ConfigurationTest` in integration test support to have a single test covering the legacy prop fidelity, then updated the pre-existing tests to use the new property.

k-wall commented:
Looking good, please address the integration test comment and address the Sonar issues.

k-wall commented on kroxylicious-integration-tests/src/test/java/io/kroxylicious/net/PassthroughProxyTest.java:
```diff
@@ -0,0 +1,43 @@
+
+    @Test
+    void testProxy() throws Exception {
+        WireMockServer wireMockServer = new WireMockServer();
```
can we use try-with-resources for the wiremockserver?

k-wall commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/clusternetworkaddressconfigprovider/SniRoutingClusterNetworkAddressConfigProvider.java:
```diff
@@ -111,43 +126,84 @@ public boolean requiresTls() {
+                throw new IllegalArgumentException("brokerAddressPattern and advertisedBrokerAddressPattern cannot both be specified");
+            }
+            if (brokerAddressPattern == null && advertisedBrokerAddressPattern == null) {
+                throw new IllegalArgumentException("brokerAddressPattern and advertisedBrokerAddressPattern both null");
```
```suggestion
                throw new IllegalArgumentException("brokerAddressPattern and advertisedBrokerAddressPattern are both null");
```

k-wall commented on kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/clusternetworkaddressconfigprovider/SniRoutingClusterNetworkAddressConfigProvider.java:
```diff
@@ -111,43 +126,84 @@ public boolean requiresTls() {
             return bootstrapAddress;
         }
 
+        public Integer getAdvertisedPort() {
```
I think the return type and the field can be `int`.
